core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java

```diff
@@ -221,15 +221,17 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
         }
         // go over and remove it from the unassigned
         for (RoutingNodes.UnassignedShards.UnassignedIterator it = routingNodes.unassigned().iterator(); it.hasNext(); ) {
-            if (it.next() != shardRouting) {
+            ShardRouting unassigned = it.next();
+            if (unassigned != shardRouting) {
                 continue;
             }
-            it.initialize(routingNode.nodeId());
-            if (shardRouting.primary()) {
-                // we need to clear the post allocation flag, since its an explicit allocation of the primary shard
-                // and we want to force allocate it (and create a new index for it)
-                routingNodes.addClearPostAllocationFlag(shardRouting.shardId());
+            // if we force allocation of a primary, we need to move the unassigned info back to treat it as if
+            // it was index creation
+            if (unassigned.primary() && unassigned.unassignedInfo().getReason() != UnassignedInfo.Reason.INDEX_CREATED) {
```

Comment:
we can't have a situation where the shard is unassigned and not have an UnassignedInfo (we assert for it on ShardRouting), and we iterate here over the unassigned shards so all should be well.